# Brandon Vasquez

### The following work was done without collaboration from other peers within the class but with aid from the SciKit library documentation page for the various methods, variables, and functions used.

## Sources used (among the documentation of the various libraries included):
* [Plot multi-class SGD on the iris dataset from SciKit Documentation](https://scikit-learn.org/stable/auto_examples/linear_model/plot_sgd_iris.html)
* [Basically what we have to do...but for the SVM classifier](https://scikit-learn.org/stable/auto_examples/svm/plot_iris_svc.html)



In [188]:
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn import datasets
from sklearn.svm import SVC
from sklearn.linear_model import SGDClassifier, LogisticRegression
from sklearn.inspection import DecisionBoundaryDisplay
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from pathlib import Path
from sklearn.metrics import accuracy_score

In [189]:
iris_data = datasets.load_iris()
features = ["Sepal Length", "Sepal Width", "Petal Length", "Petal Width"]

match_targets = np.array([iris_data.target_names[target] for target in iris_data.target]) # match each target name to the respective target index
iris_df = pd.DataFrame(iris_data.data, columns=features, index=match_targets)

compiled_accurracies = []
print(type(iris_df), iris_df.head(10))
print(match_targets, len(match_targets))

<class 'pandas.core.frame.DataFrame'>         Sepal Length  Sepal Width  Petal Length  Petal Width
setosa           5.1          3.5           1.4          0.2
setosa           4.9          3.0           1.4          0.2
setosa           4.7          3.2           1.3          0.2
setosa           4.6          3.1           1.5          0.2
setosa           5.0          3.6           1.4          0.2
setosa           5.4          3.9           1.7          0.4
setosa           4.6          3.4           1.4          0.3
setosa           5.0          3.4           1.5          0.2
setosa           4.4          2.9           1.4          0.2
setosa           4.9          3.1           1.5          0.1
['setosa' 'setosa' 'setosa' 'setosa' 'setosa' 'setosa' 'setosa' 'setosa'
 'setosa' 'setosa' 'setosa' 'setosa' 'setosa' 'setosa' 'setosa' 'setosa'
 'setosa' 'setosa' 'setosa' 'setosa' 'setosa' 'setosa' 'setosa' 'setosa'
 'setosa' 'setosa' 'setosa' 'setosa' 'setosa' 'setosa' 'setosa' 'setosa'

In [190]:
# helper functions here....
def addScore(clf: str, test_s: float, train_s: float) -> None:
    compiled_accurracies.append({
        'clf': clf,
        'test_accu': test_s,
        'train_accu': train_s
    })
    return None

def implementation_of_score_func(*arrays) -> float: #just for fun implementing the score function for SGD using predict
    out = [1 if predict == arrays[-1][i] else 0 for i, predict in enumerate(arrays[0])]
    return out.count(1) / len(out)

In [191]:
sep_feats = iris_df[['Sepal Length', 'Sepal Width']] # plot this
print(type(sep_feats), sep_feats[::10])

<class 'pandas.core.frame.DataFrame'>             Sepal Length  Sepal Width
setosa               5.1          3.5
setosa               5.4          3.7
setosa               5.4          3.4
setosa               4.8          3.1
setosa               5.0          3.5
versicolor           7.0          3.2
versicolor           5.0          2.0
versicolor           5.9          3.2
versicolor           5.5          2.4
versicolor           5.5          2.6
virginica            6.3          3.3
virginica            6.5          3.2
virginica            6.9          3.2
virginica            7.4          2.8
virginica            6.7          3.1


In [192]:
X_train,X_test,y_train,y_test = train_test_split(sep_feats, match_targets, test_size=0.2, random_state=805)

In [193]:
print(X_train.shape, X_train[::5], '\n\nTraining features\n\n') # plot this
print(y_train.shape, y_train[::5], '\n\nTraining labels')

(120, 2)             Sepal Length  Sepal Width
versicolor           4.9          2.4
virginica            6.8          3.2
setosa               5.1          3.5
setosa               5.3          3.7
virginica            6.0          3.0
setosa               4.9          3.6
setosa               4.7          3.2
virginica            6.7          3.0
versicolor           5.6          3.0
setosa               5.1          3.8
versicolor           6.8          2.8
virginica            4.9          2.5
versicolor           6.4          2.9
virginica            6.5          3.0
virginica            5.7          2.5
virginica            7.1          3.0
setosa               5.2          3.4
versicolor           6.3          2.3
versicolor           5.8          2.7
virginica            6.9          3.1
virginica            7.3          2.9
versicolor           5.0          2.0
setosa               4.6          3.1
versicolor           5.7          2.9 

Training features


(120,) ['versicolor

In [194]:
print(X_test.shape, X_test[::2], '\n\nTraining features\n\n') # plot this
print(y_test.shape, y_test[::2], '\n\nTraining labels')

(30, 2)             Sepal Length  Sepal Width
versicolor           5.6          2.7
versicolor           6.6          3.0
setosa               4.4          2.9
versicolor           5.5          2.6
virginica            7.7          3.8
setosa               5.4          3.9
setosa               5.1          3.8
virginica            7.2          3.2
setosa               5.7          4.4
setosa               5.1          3.7
versicolor           5.8          2.7
virginica            7.2          3.6
virginica            6.4          2.7
setosa               5.0          3.6
virginica            7.4          2.8 

Training features


(30,) ['versicolor' 'versicolor' 'setosa' 'versicolor' 'virginica' 'setosa'
 'setosa' 'virginica' 'setosa' 'setosa' 'versicolor' 'virginica'
 'virginica' 'setosa' 'virginica'] 

Training labels


In [195]:
#SGD with hinge loss function
clf_hinge = SGDClassifier(loss="hinge", tol=1e-4, random_state=0)
clf_hinge = clf_hinge.fit(X_train, y_train)


print(accuracy_score(y_train, clf_hinge.predict(X_train)))
print(accuracy_score(y_test, clf_hinge.predict(X_test)))
print(f'Testing {round(clf_hinge.score(X_test, y_test), 3) * 100}\nTraining: {round(clf_hinge.score(X_train, y_train), 3) * 100}')

addScore('SGD_hinge', round(clf_hinge.score(X_test, y_test), 3) * 100, round(clf_hinge.score(X_train, y_train), 3) * 100)

0.7166666666666667
0.7
Testing 70.0
Training: 71.7


In [196]:
#SGD with modified huber loss function
clf_mod_huber = SGDClassifier(loss="modified_huber", tol=1e-4, random_state=0)
clf_mod_huber = clf_mod_huber.fit(X_train, y_train)

print(accuracy_score(y_train, clf_mod_huber.predict(X_train)))
print(accuracy_score(y_test, clf_mod_huber.predict(X_test)))
print(f'Testing {round(clf_mod_huber.score(X_test, y_test), 3) * 100}\nTraining: {round(clf_mod_huber.score(X_train, y_train), 3) * 100}')

addScore('SGD_mod_huber', round(clf_mod_huber.score(X_test, y_test), 3), round(clf_mod_huber.score(X_train, y_train), 3) * 100)

0.5583333333333333
0.6666666666666666
Testing 66.7
Training: 55.800000000000004


In [197]:
logist_clf = LogisticRegression(penalty='l2', solver='newton-cg', random_state=0)
logist_clf = logist_clf.fit(X_train, y_train)

print(accuracy_score(y_train, logist_clf.predict(X_train)))
print(accuracy_score(y_test, logist_clf.predict(X_test)))
print(f'Testing {round(logist_clf.score(X_test, y_test), 3) * 100}\nTraining: {round(logist_clf.score(X_train, y_train), 3) * 100}')

addScore('Logistic Regression', round(logist_clf.score(X_test, y_test), 3), round(logist_clf.score(X_train, y_train), 3) * 100)

0.825
0.7333333333333333
Testing 73.3
Training: 82.5


In [198]:
print(compiled_accurracies)

[{'clf': 'SGD_hinge', 'test_accu': 70.0, 'train_accu': 71.7}, {'clf': 'SGD_mod_huber', 'test_accu': 0.667, 'train_accu': 55.800000000000004}, {'clf': 'Logistic Regression', 'test_accu': 0.733, 'train_accu': 82.5}]


In [199]:
SVM_linker_clf = SVC(kernel='linear')
SVM_linker_clf = SVM_linker_clf.fit(X_train, y_train)

print(accuracy_score(y_train, SVM_linker_clf.predict(X_train)))
print(accuracy_score(y_test, SVM_linker_clf.predict(X_test)))
print(f'Testing {round(SVM_linker_clf.score(X_test, y_test), 3) * 100}\nTraining: {round(SVM_linker_clf.score(X_train, y_train), 3) * 100}')

addScore('SVM_linker_clf', round(SVM_linker_clf.score(X_test, y_test), 3), round(SVM_linker_clf.score(X_train, y_train), 3) * 100)

0.8333333333333334
0.7666666666666667
Testing 76.7
Training: 83.3


In [200]:
SVM_rbfker= SVC(kernel='rbf', gamma='auto') #gamma of auto allows it to use a 1 / n_features
SVM_rbfker = SVM_rbfker.fit(X_train, y_train)

print(accuracy_score(y_train, SVM_rbfker.predict(X_train)))
print(accuracy_score(y_test, SVM_rbfker.predict(X_test)))
print(f'Testing {round(SVM_rbfker.score(X_test, y_test), 3) * 100}\nTraining: {round(SVM_rbfker.score(X_train, y_train), 3) * 100}')

addScore('SVM_rbfker', round(SVM_rbfker.score(X_test, y_test), 3), round(SVM_rbfker.score(X_train, y_train), 3) * 100)
SVM_rbfker.get_params()

0.8333333333333334
0.7333333333333333
Testing 73.3
Training: 83.3


{'C': 1.0,
 'break_ties': False,
 'cache_size': 200,
 'class_weight': None,
 'coef0': 0.0,
 'decision_function_shape': 'ovr',
 'degree': 3,
 'gamma': 'auto',
 'kernel': 'rbf',
 'max_iter': -1,
 'probability': False,
 'random_state': None,
 'shrinking': True,
 'tol': 0.001,
 'verbose': False}

In [201]:
dec_tree_clf = DecisionTreeClassifier(criterion='gini')
dec_tree_clf = dec_tree_clf.fit(X_train, y_train)

print(accuracy_score(y_train, dec_tree_clf.predict(X_train)))
print(accuracy_score(y_test, dec_tree_clf.predict(X_test)))
print(f'Testing {round(SVM_rbfker.score(X_test, y_test), 3) * 100}\nTraining: {round(dec_tree_clf.score(X_train, y_train), 3) * 100}')

addScore('dec_tree_clf', round(dec_tree_clf.score(X_test, y_test), 3), round(dec_tree_clf.score(X_train, y_train), 3) * 100)
dec_tree_clf.get_params()

0.9333333333333333
0.7333333333333333
Testing 73.3
Training: 93.30000000000001


{'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'monotonic_cst': None,
 'random_state': None,
 'splitter': 'best'}

In [202]:
rf_clf = RandomForestClassifier(criterion='log_loss', n_estimators=200).fit(X_train, y_train)

print(accuracy_score(y_train, rf_clf.predict(X_train)))
print(accuracy_score(y_test, rf_clf.predict(X_test)))
print(f'Testing {round(rf_clf.score(X_test, y_test), 3) * 100}\nTraining: {round(rf_clf.score(X_train, y_train), 3) * 100}')

addScore('rf_clf', round(rf_clf.score(X_test, y_test), 3), round(rf_clf.score(X_train, y_train), 3) * 100)
rf_clf.get_params()

0.9333333333333333
0.7333333333333333
Testing 73.3
Training: 93.30000000000001


{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'log_loss',
 'max_depth': None,
 'max_features': 'sqrt',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'monotonic_cst': None,
 'n_estimators': 200,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

In [203]:
knn_clf_k47 = KNeighborsClassifier(n_neighbors=47).fit(X_train, y_train)

print(accuracy_score(y_train, knn_clf_k47.predict(X_train)))
print(accuracy_score(y_test, knn_clf_k47.predict(X_test)))
print(f'Testing {round(knn_clf_k47.score(X_test, y_test), 3) * 100}\nTraining: {round(knn_clf_k47.score(X_train, y_train), 3) * 100}')

addScore('knn_clf_k47', round(knn_clf_k47.score(X_test, y_test), 3), round(knn_clf_k47.score(X_train, y_train), 3) * 100)
knn_clf_k47.get_params()

0.8083333333333333
0.7666666666666667
Testing 76.7
Training: 80.80000000000001


{'algorithm': 'auto',
 'leaf_size': 30,
 'metric': 'minkowski',
 'metric_params': None,
 'n_jobs': None,
 'n_neighbors': 47,
 'p': 2,
 'weights': 'uniform'}

In [204]:
knn_clf_k3 = KNeighborsClassifier(n_neighbors=3).fit(X_train, y_train)

print(accuracy_score(y_train, knn_clf_k3.predict(X_train)))
print(accuracy_score(y_test, knn_clf_k3.predict(X_test)))
print(f'Testing {round(knn_clf_k3.score(X_test, y_test), 3) * 100}\nTraining: {round(knn_clf_k3.score(X_train, y_train), 3) * 100}')

addScore('knn_clf_k3', round(knn_clf_k3.score(X_test, y_test), 3), round(knn_clf_k3.score(X_train, y_train), 3) * 100)
knn_clf_k3.get_params()

0.8416666666666667
0.7666666666666667
Testing 76.7
Training: 84.2


{'algorithm': 'auto',
 'leaf_size': 30,
 'metric': 'minkowski',
 'metric_params': None,
 'n_jobs': None,
 'n_neighbors': 3,
 'p': 2,
 'weights': 'uniform'}

In [206]:
print(compiled_accurracies)
print(len(compiled_accurracies))

[{'clf': 'SGD_hinge', 'test_accu': 70.0, 'train_accu': 71.7}, {'clf': 'SGD_mod_huber', 'test_accu': 0.667, 'train_accu': 55.800000000000004}, {'clf': 'Logistic Regression', 'test_accu': 0.733, 'train_accu': 82.5}, {'clf': 'SVM_linker_clf', 'test_accu': 0.767, 'train_accu': 83.3}, {'clf': 'SVM_rbfker', 'test_accu': 0.733, 'train_accu': 83.3}, {'clf': 'dec_tree_clf', 'test_accu': 0.733, 'train_accu': 93.30000000000001}, {'clf': 'rf_clf', 'test_accu': 0.733, 'train_accu': 93.30000000000001}, {'clf': 'knn_clf_k47', 'test_accu': 0.767, 'train_accu': 80.80000000000001}, {'clf': 'knn_clf_k3', 'test_accu': 0.767, 'train_accu': 84.2}]
9


# Discussion

## Large Delta of two feature classifications when using Perceptron loss function
* Choosing two distinct sets of features it was observed that the loss functions effect on the classifying the data drastically changed as observed in the 'Scores across the various # of features' it can be seen that when we compare the accurracy of the perceptron loss function on the two different sets of features this was where the greatest disparity was seen. When now reflecting on the plot featuring the decision regions of the respective classes we see that for the set consisting of the features 'Petal Width, Sepal Width' there was a gross misclassification for the Versicolor class where the plotted training points extend into the decision regions--not to be confused with the hyperplane--of both setosa and virginica. Now comparing this classification plot to that of the set with 'Petal Width, Petal Length' it can be observed that the higher score of the classification is due to the more accurate split of the decision regions where the training data--while still not completely linearly seperable. This delta between the two further shows the effect that linearly inseperable data has on the percetron loss function and thus the classifier as a whole.

## Accurracy Across the chosen combination of features...
* When looking across all of the features in the plot 'Scores across the various combination of features' tt is also noted that using the perceptron loss function on the three-feature combination as well as the two distinct two-feature combinations resulted in the lowest accuracy score, however, the four-feature combination exhibited the second highest accuracy score with the perceptron loss function. The highest accuracy score recorded using the four loss functions for the ‘Sepal Width, Petal Width’ two-feature and four-feature combination was using the hinge loss function meanwhile for the ‘Petal Width, Petal Length’ and the three-feature loss function the highest accuracy score was recorded with the squared hinge loss function. 

* Inclusively the 'Petal Width, Sepal Width' two-feature combination exhibited similar deltas in accurracy and even slightly converged on the loss function that resulted in the highest accuracy score for both--squared hinge--this could have possibly been due to the similarity in the distribution of the training data, when observing the 'loss function: squared_hinge with score 0.933' from the 3D plot of the three feature classification combination and the 'loss function: squared_hinge with score 0.933' in 2D featuring the 'Petal Width, Sepal Width' two-feature combination it can be observed that the distribution of the three species is quite similar with the setosa class being linearly seperable from the other two class meanwhile versicolor and virginica are not linearly seperable.







